Fix s3 requests without region

vitlibar opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry:
Use `GetObjectAttributes` request instead of `HeadObject` request to get the size of an object in AWS S3.
This change fixes handling endpoints without explicit region, for example

```
select * from s3('https://test.s3.amazonaws.com/mydata.csv')
```

vitlibar committed:
Stop using HeadObject requests in S3
because they don't work well with endpoints without explicit region.

vitlibar committed:
Move `throw_on_error` parameter to the end.

antonio2368 commented on src/IO/S3Common.cpp:
```diff
@@ -894,61 +919,84 @@ namespace S3
+    std::pair<bool /* exists */, Aws::S3::S3Error> checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id, bool for_disk_s3)
+    {
+        auto outcome = getObjectAttributes(client, bucket, key, version_id, for_disk_s3);
+        if (outcome.IsSuccess())
+            return {true, {}};
```
Maybe return `{}` for error if it's NotFound?

vitlibar replied:
This function is called when we're quite sure that an object must exist. So we need to show an error if it's not.

vitlibar replied:
I changed the code so now the function just throws an exception if the object doesn't exist without returning anything.

vitlibar commented on src/IO/S3Common.cpp:
```diff
@@ -894,61 +919,84 @@ namespace S3
 
-        const auto & error = outcome.GetError();
-        if (isNotFoundError(error.GetErrorType()))
+        if (!throw_on_error || isNotFoundError(error.GetErrorType()))
```
I'll remove `throw_on_error` parameter here, we don't really need it anyway.

vitlibar committed:
Make checkObjectExists() easier.

antonio2368 commented on src/IO/S3Common.h:
```diff
@@ -121,21 +121,35 @@ struct URI
 
-S3::ObjectInfo getObjectInfo(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id, bool throw_on_error, bool for_disk_s3);
+/// Throws an exception if a specified object doesn't exist. `description` is used as a part of the error message.
+void checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false, const std::string_view & description = {});
```
```suggestion
void checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false, std::string_view description = {});
```

vitlibar replied:
ok

vitlibar committed:
Use `std::string_view` instead of `const std::string_view &`

vitlibar committed:
Fix build.

vitlibar commented:
This is bad. It appears Minio doesn't support `GetObjectAttributes` requests, that's why many tests failed.

vitlibar committed:
More complex logic: GetObjectAttributes requests will be used
only if the endpoint is "*.amazonaws.com", otherwise HeadObject requests will be used.
